[CMake] fix symlink creation on Windows

hughbe opened the pull request:
Introduce `swift_create_post_build_symlink` and a batch script to delete an existing symlink and create a new one.

The reason I need a batch script is because CMake's `if(EXISTS)` doesn't work for symlinks, CMakes `create_symlink` doesn't work on Windows, and `mklink` throws errors when creating a symlink when a symlink already exists.

I can't use `swift_create_post_build_symlink`  in `stdlib/public/SwiftShims/CMakeLists.txt` because this is a target, not a command

@compnerd @jrose-apple

compnerd commented:
CC @llvm-beanz

I think that Chris had some fixes for the llvm macros and we should try to reuse those.

hughbe commented:
> I think that Chris had some fixes for the llvm macros and we should try to reuse those.

I thought these were for executables? Not .py and .dylib files

llvm-beanz commented on lib/SwiftDemangle/CMakeLists.txt:
```diff
@@ -9,14 +9,12 @@ swift_install_in_component(compiler
+swift_create_post_build_symlink(swiftDemangle
+  FALSE
+  "libswiftDemangle.dylib"
+  "libfunctionNameDemangle.dylib"
```
Is this even needed for Windows? This is really a hack to workaround a library being renamed. I'm not even sure we actually need to create this symlink at all anymore.

hughbe replied:
@jrose-apple I think you added this back in 2015. What do you think?

llvm-beanz commented on stdlib/public/SwiftShims/CMakeLists.txt:
```diff
@@ -79,7 +79,20 @@ else() # NOT SWIFT_BUILT_STANDALONE
+    COMMAND
+      "${CMAKE_COMMAND}" "-E" "make_directory" "${SWIFTLIB_DIR}"
+    COMMAND
+      "mklink" "/J"
```
In LLVM our normal approach is to replace "create_symlink" with "copy" on non-unix systems. I think that is way simpler, and a better approach. Especially because calling `mklink` on Windows requires elevated privileges, and I think it would be unfortunate if building Swift required elevated privileges. That would specifically hamper the ability of students to checkout and build Swift on University-owned machines.

hughbe replied:
Hmm,  have an account with elevated privileges. I've noticed that I get privilege errors using `mklink` Without any arguments. They seem to go away when I add `/H` or `/J`.

That said, i can change to be a basic copy_if_different/copy if you'd like.

hughbe replied:
Aye aye, will do - thanks

hughbe commented:
Thanks @llvm-beanz - I've updated the PR to use copying instead of symlinking on Windows. In the second commit I removed the demangle symlinking as you suggested

llvm-beanz commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -123,3 +123,23 @@ function(set_with_default variable value)
   endif()
 endfunction()
+
+function(swift_create_post_build_symlink target is_directory path_to_file path_to_symlink working_directory comment)
```
Rather than lots of named parameters, why not use `cmake_parse_arguments` with named parameters? That is much easier to read.

I'm fine with the source, dest, and target being named parameters, but `is_directory` should be a flag, and `working_directory` and `comment` should be single option parameters (potentially even optional).

hughbe commented:
@llvm-beanz  I've addressed your comments, looks nicer now

llvm-beanz commented:
@hughbe LGTM! Thanks!

slavapestov commented:
@swift-ci Please smoke test
